File: engine/publishers/medium.py
"""
medium.py — Medium API publisher.
Uses Medium API v1: https://github.com/Medium/medium-api-docs
"""

import requests
from publishers.base import BasePublisher, PublishResult


class MediumPublisher(BasePublisher):
    """Publish to Medium via their official API."""

    def __init__(self):
        super().__init__("medium")

    def check_auth(self):
        try:
            self._require_credential("token")
            return True
        except ValueError:
            return False

    def _get_author_id(self, token):
        """Get Medium user ID."""
        resp = requests.get(
            "https://api.medium.com/v1/me",
            headers={"Authorization": f"Bearer {token}",
                     "Content-Type": "application/json",
                     "Accept": "application/json"},
            timeout=30
        )
        if resp.status_code == 200:
            return resp.json()["data"]["id"]
        return None

    def publish(self, title, body, content_type="article", options=None):
        opts = options or {}
        token = self._get_credential("token")
        author_id = self._get_credential("author_id") or self._get_author_id(token)

        if not author_id:
            return PublishResult.fail("Could not get Medium author ID. Check token.")

        url = f"https://api.medium.com/v1/users/{author_id}/posts"

        # Format body as Medium markdown if it's HTML
        content_format = opts.get("format", "html")
        tags = opts.get("tags", [])
        canonical_url = opts.get("canonical_url", "")
        publish_status = opts.get("status", "draft")  # draft, unlisted, public

        data = {
            "title": title,
            "contentFormat": content_format,
            "content": body,
            "tags": tags[:5],  # Max 5 tags
            "publishStatus": publish_status,
        }
        if canonical_url:
            data["canonicalUrl"] = canonical_url

        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

        try:
            resp = requests.post(url, json=data, headers=headers, timeout=60)
            if resp.status_code in (200, 201):
                result = resp.json()["data"]
                return PublishResult.ok(
                    url=result.get("url", ""),
                    post_id=result.get("id", "")
                )
            else:
                err = resp.json().get("errors", [{}])[0].get("message", resp.text)[:300]
                return PublishResult.fail(f"HTTP {resp.status_code}: {err}", resp.status_code)

        except requests.exceptions.RequestException as e:
            return PublishResult.fail(f"Request failed: {str(e)[:200]}")

File: src/lib/publishers/medium.ts
// Medium Publisher — Medium REST API publisher
// Publishes content to Medium via the REST API using an integration token
// API Docs: https://github.com/Medium/medium-api-docs

import { type PublishPayload, type PublishResult, type Publisher } from "./index";

class MediumPublisher implements Publisher {
  name = "Medium";
  platform = "medium" as const;

  async validate(credentials: Record<string, string>): Promise<boolean> {
    const { token } = credentials;

    if (!token) return false;

    try {
      const response = await fetch("https://api.medium.com/v1/me", {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      return response.ok;
    } catch {
      return false;
    }
  }

  async publish(
    payload: PublishPayload,
    credentials: Record<string, string>
  ): Promise<PublishResult> {
    const { token, endpointUrl } = credentials;

    if (!token) {
      return {
        success: false,
        platform: "medium",
        error: "Medium integration token is required",
      };
    }

    try {
      // First, get the authenticated user's ID
      const meResponse = await fetch("https://api.medium.com/v1/me", {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });

      if (!meResponse.ok) {
        const errorText = await meResponse.text();
        return {
          success: false,
          platform: "medium",
          error: `Medium authentication failed: ${errorText}`,
        };
      }

      const meData = await meResponse.json();
      const authorId = meData?.data?.id;

      if (!authorId) {
        return {
          success: false,
          platform: "medium",
          error: "Could not retrieve Medium user ID",
        };
      }

      // Determine the publish status
      // Medium supports: public, draft, unlisted
      const publishStatus =
        payload.status === "published" ? "public" : "draft";

      // Determine content format — default to markdown
      const contentFormat = "markdown";

      // Build the request body
      const body: Record<string, unknown> = {
        title: payload.title,
        contentFormat,
        content: payload.body,
        publishStatus,
      };

      // Add tags if provided (Medium allows up to 5 tags)
      if (payload.tags && payload.tags.length > 0) {
        body.tags = payload.tags.slice(0, 5);
      }

      // If a publication ID is provided via endpointUrl, publish to that publication
      // Otherwise, publish under the user's own profile
      let postUrl: string;
      if (endpointUrl && endpointUrl.trim() !== "") {
        // endpointUrl stores the publicationId
        postUrl = `https://api.medium.com/v1/publications/${endpointUrl}/posts`;
      } else {
        postUrl = `https://api.medium.com/v1/me/${authorId}/posts`;
      }

      const response = await fetch(postUrl, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        const errorText = await response.text();
        return {
          success: false,
          platform: "medium",
          error: `Medium API error: ${errorText}`,
        };
      }

      const data = await response.json();

      return {
        success: true,
        platform: "medium",
        postId: data?.data?.id ? String(data.data.id) : undefined,
        postUrl: data?.data?.url || undefined,
        responsePayload: data,
      };
    } catch (error) {
      return {
        success: false,
        platform: "medium",
        error: error instanceof Error ? error.message : "Medium publish failed",
      };
    }
  }
}

export default new MediumPublisher();
